### PR TITLE
fix(app): mag block with staging area location conflict, handle no configurable modules

### PR DESCRIPTION
**Pull request by b-cooper**
# Overview

Add special-case affordances for the magnetic block with staging area slot fixture in the protocol
setup for run experience on the ODD and desktop. Show different modals depending on the presence of
matching but configured modules in the ChooseModuleToConfigure conflict resolution experience on ODD
and desktop

Closes [PLAT-298](https://opentrons.atlassian.net/browse/PLAT-298), Closes [PLAT-293](https://opentrons.atlassian.net/browse/PLAT-293), Closes [PLAT-294](https://opentrons.atlassian.net/browse/PLAT-294)
Closes RQA-2670
# Review requests

- Start protocol setup with a protocol that specs the mag block in a slot adjacent to a column 4 slot that is also used in the protocol. Confirm ODD and desktop setup items render properly and location conflict modal resolves as expected 

- Start protocol setup for a protocol that uses a module that is not currently configured or attached. 
  - Confirm that ODD and desktop location conflict prompts to plug in and turn on module with new grey loading state and no button
  
- Start protocol setup for a protocol that uses a module that is configured into a different location than the protocol expects.
  - Confirm the the ODD location conflict provides button to cancel run and exit to the Protocol List page
  - Confirm the desktop location conflict provides button to navigate back to the device detail page and repair deck config manually

# Risk assessment
low

[PLAT-298]: https://opentrons.atlassian.net/browse/PLAT-298?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[PLAT-293]: https://opentrons.atlassian.net/browse/PLAT-293?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[PLAT-294]: https://opentrons.atlassian.net/browse/PLAT-294?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: app/src/pages/Protocols/hooks/index.ts [hunk 1]
```diff
@@ -65,6 +68,9 @@ export type ProtocolHardware =
   | ProtocolFixture
 
 const DECK_CONFIG_REFETCH_INTERVAL = 5000
+const USB_MODULE_ADDRESSABLE_AREAS = FLEX_MODULE_ADDRESSABLE_AREAS.filter(
```
Comment (brenthagen) [inline app/src/pages/Protocols/hooks/index.ts]: can use `FLEX_USB_MODULE_ADDRESSABLE_AREAS` constant directly?

File: app/src/organisms/Devices/ProtocolRun/SetupModuleAndDeck/ChooseModuleToConfigureModal.tsx [hunk 2]
```diff
@@ -195,3 +185,86 @@ export const ChooseModuleToConfigureModal = (
...
+interface NoUnconfiguredModulesProps {
+  moduleDisplayName: string
+  displaySlotName: string
+  configuredModuleMatches: AttachedModule[]
+  isOnDevice: boolean
+  robotName: string
+}
+function NoUnconfiguredModules(props: NoUnconfiguredModulesProps): JSX.Element {
+  const { moduleDisplayName, displaySlotName, isOnDevice, robotName } = props
+  const configuredModuleMatches = ['feer']
```
Comment (brenthagen) [inline app/src/organisms/Devices/ProtocolRun/SetupModuleAndDeck/ChooseModuleToConfigureModal.tsx]: is this WIP?

File: app/src/organisms/Devices/ProtocolRun/SetupModuleAndDeck/ChooseModuleToConfigureModal.tsx [hunk 3]
```diff
@@ -195,3 +185,86 @@ export const ChooseModuleToConfigureModal = (
...
+        </>
+      ) : (
+        <>
+          <StyledText as="p">
+            {t('there_are_no_unconfigured_modules', {
+              module: moduleDisplayName,
+              slot: displaySlotName,
+            })}
+          </StyledText>
+          {loadingBlock}
```
Comment (brenthagen) [inline app/src/organisms/Devices/ProtocolRun/SetupModuleAndDeck/ChooseModuleToConfigureModal.tsx:270]: is no exit button on this version intentional?

Screen Shot 2024-04-30 at 3 35 56 PM

File: app/src/organisms/ProtocolDetails/index.tsx [hunk 4]
```diff
@@ -236,7 +238,13 @@ export function ProtocolDetails(
...
   const requiredModuleDetails =
     mostRecentAnalysis?.commands != null
-      ? map(parseInitialLoadedModulesBySlot(mostRecentAnalysis.commands))
+      ? map(
+          parseInitialLoadedModulesBySlot(mostRecentAnalysis.commands)
+        ).filter(
+          loadedModule =>
+            // filter out magnetic block which is already handled by the required fixture details
+            getModuleType(loadedModule.params.model) !== MAGNETIC_BLOCK_TYPE
+        )
```
Comment (brenthagen) [inline app/src/organisms/ProtocolDetails/index.tsx:247]: need to filter out the other modules also, they are rendering twice in the `RobotConfigurationDetails` list

Screen Shot 2024-04-30 at 3 57 49 PM

Comment (brenthagen) [reply app/src/organisms/ProtocolDetails/index.tsx:247]: or maybe filter all modules out of `requiredFixtureDetails` instead

File: app/src/pages/ProtocolDetails/Hardware.tsx [hunk 5]
```diff
@@ -94,7 +96,16 @@ const useHardwareName = (protocolHardware: ProtocolHardware): string => {
...
-    return getFixtureDisplayName(protocolHardware.cutoutFixtureId)
+    let displayFixtureId = protocolHardware.cutoutFixtureId
+    // Only show staging area as a requirement for the staging area with mag block fixture,
+    // as mag block requirement is handled as a part of modules, we don't want to list it twice
+    if (
+      protocolHardware.cutoutFixtureId ===
+      STAGING_AREA_SLOT_WITH_MAGNETIC_BLOCK_V1_FIXTURE
+    ) {
+      displayFixtureId = STAGING_AREA_RIGHT_SLOT_FIXTURE
+    }
```
Comment (brenthagen) [inline app/src/pages/ProtocolDetails/Hardware.tsx]: discussed IRL, don't want to assign `STAGING_AREA_RIGHT_SLOT_FIXTURE` here

File: app/src/organisms/Devices/ProtocolRun/SetupModuleAndDeck/ChooseModuleToConfigureModal.tsx (hunk 3 above)
Comment (b-cooper) [reply app/src/organisms/Devices/ProtocolRun/SetupModuleAndDeck/ChooseModuleToConfigureModal.tsx:270]: That's correct, the exit button in the top left is the only action the user can take in this case (other than plugging in a module of the specified model)

**Review by brenthagen — APPROVED** [review]
LGTM !

**Pull request merged**